# Introductory guide

The basics of SVX.

## Prerequisites

- Download the [latest SVX binary](https://github.com/PacificBiosciences/svx/releases) or [clone](https://github.com/PacificBiosciences/svx) and build using `cargo build --release` (Rust >=v1.85.0)
- At least two VCF files generated by [sawfish](https://github.com/PacificBiosciences/sawfish)
- (Optional): A tandem repeat definition BED file such as referred to on the [TRGT](https://github.com/PacificBiosciences/trgt?tab=readme-ov-file#repeat-catalogs) repository 

## Running SVX

SVX can be run from the CLI to merge VCFs with structural variant calls, for example if we have two VCFs: `data/vcf_1.vcf.gz` and `data/vcf_2.vcf.gz`, we run:

```
svx merge --vcf data/vcf_1.vcf.gz data/vcf_2.vcf.gz
```

This will use default settings everywhere and write the merged VCF to stdout in an uncompressed format. Note that all VCFs in a directory may be given as `--vcf data/*.vcf.gz` or as a file of filenames, i.e., `--vcf-list`. To write to a file, we can add `-o out.vcf` for uncompressed or `-o out.vcf.gz` as compressed (the `--output-type` flag may be used to define overrides for this behavior).

To enable SV annotation by user provided TR definitions and special handling of such annotated SVs add the `--trs tr.bed` flag, further customization is possible with [other CLI flags](cli.md), which allows changing settings to influence the merging process.

SVX (currently) parallelizes on the contig and svtype level (`-@` flag), this can significantly speed-up the merging process. For testing, it can be useful to first limit the scope of SV merging. Scope may be constrained on the contig level using the `--contig` flag, for example `--contig chr1` will only analyze events on chromosome 1. However this can also be set in a more granular manner using the `--target-positions` flag where any number of ranges can be provided as `--target-positions "(chr1:50000-100000),(chr5:100000-200000)"` to only analyze SVs with breakpoints within these ranges and/or single positions such as ``--target-positions "(chr1:100000),(chr6:55000)"` such that only SVs at the exact breakpoints are processed.
